hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockManager.java

```diff
@@ -3559,9 +3558,26 @@ private Block addStoredBlock(final BlockInfo block,
     if ((corruptReplicasCount > 0) && (numLiveReplicas >= fileRedundancy)) {
       invalidateCorruptReplicas(storedBlock, reportedBlock, num);
     }
+    if (shouldInvalidateDecommissionedRedundancy(num, fileRedundancy)) {
+      for (DatanodeStorageInfo storage : blocksMap.getStorages(block)) {
+        final DatanodeDescriptor datanode = storage.getDatanodeDescriptor();
+        if (datanode.isDecommissioned()
+            || datanode.isDecommissionInProgress()) {
+          addToInvalidates(storedBlock, datanode);
+        }
+      }
+    }
     return storedBlock;
   }
 
+  // If there are enough live replicas, start invalidating
+  // decommissioned + decommissioning replicas
+  private boolean shouldInvalidateDecommissionedRedundancy(NumberReplicas num,
```

Comment:
It makes sense. Maybe we should describe some of the JIRA description in this method to explain what we are doing in the high level.

Reply:
Good idea.